[BugFix] SyncDataCollector init when device and env_device are different

albertbou92 opened the pull request:
## Description

In the init method of the SyncDataCollector class, a small number of steps is taken with the policy to determine the relevant keys of the output TensorDict. When the policy device and the environment device are different, that can raise a RuntimeError since the input provided to the policy is located in the environment device.

This PR only makes sure that the TensorDict provided to the policy is in the policy device, and then moves the output TensorDict to the environment device again. 

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

vmoens reviewed (commented):
Thanks for this! Can you add a test where device and passing_device differ to check non-regression over this bug fix?

albertbou92 commented:
Added tests for all device/passing_device combinations for SyncDataCollector, MultiSyncDataCollector and MultiaSyncDataCollector. 

I noticed that for MultiSyncDataCollector, the output batch TensorDict has its device attribute set to None.
I think the problem can be in the TensorDict.cat method, which returns a TensorDict whose "next" TensorDict has device set to None. As a sanity check, I added a line of code to make sure the MultiSyncDataCollector output TensorDict is cast to the correct device.

vmoens commented:
@albertbou92 interesting
> I noticed that for MultiSyncDataCollector, the yielded TensorDicts had its device attribute set to None.

Does this happen when the passing devices match too? Or just when there's more than one?
`torch.cat` should conserve the device (since we're sure that all device match). If it doesn't it's a bug. When the devices don't match, we first cast things to cpu before calling cat. It's definitely not ultra fast so we may want to parametrise that feature.

> which returns a TensorDict whose "next" TensorDict has device set to None.
You mean that the tensordict has a device but the nested tensordict doesn't? Weird...

albertbou92 commented:
> @albertbou92 interesting
> 
> > I noticed that for MultiSyncDataCollector, the yielded TensorDicts had its device attribute set to None.
> 
> Does this happen when the passing devices match too? Or just when there's more than one? `torch.cat` should conserve the device (since we're sure that all device match). If it doesn't it's a bug. When the devices don't match, we first cast things to cpu before calling cat. It's definitely not ultra fast so we may want to parametrise that feature.
> 
> > which returns a TensorDict whose "next" TensorDict has device set to None.
> > You mean that the tensordict has a device but the nested tensordict doesn't? Weird...

albertbou92 commented:
> Does this happen when the passing devices match too?

Yes it happens for all device combinations

albertbou92 commented:
> You mean that the tensordict has a device but the nested tensordict doesn't? Weird...

exactly

vmoens commented:
> > You mean that the tensordict has a device but the nested tensordict doesn't? Weird...
> 
> 
> 
> exactly

Ok then it's a tensordict bug, I'll have a look into it.

albertbou92 commented:
Should I remove the line of code to make sure the MultiSyncDataCollector yielded TensorDicts are cast to the correct device.

We could keep it as a sanity check, won't add overhead if the TensorDicts is already in the correct device.

vmoens commented:
> Should I remove the line of code that makes sure the MultiSyncDataCollector yielded TensorDicts are cast to the correct device.
> 
> We could keep it as a sanity check, won't add overhead if the TensorDicts is already in the correct device.

Which line of code are we talking about?

By the way: can you check if the device bug with nested tensordicts is still present?

albertbou92 commented:
the device bug with nested TensorDicts seems to be fixed now!
 
In my code I added to the MultiSyncDataCollector _iterator_ method a `out_buffer = out_buffer.to(prev_device)` after concatenating all different dicts to make sure the yielded TensorDict was places to the passing_device. Now that the cat method is fixed might be unnecessary. 
 
I can remove it or we can either keep as a sanity check.

vmoens commented:
If it's not necessary I would remove it. Those things take little time but they quickly pile up to a consequent overhead.

albertbou92 commented:
ok, removed that checking line.

Now the PR just fixed the bug in  SyncDataCollector when policy device and passing_device are different and adds tests to make sure all collectors work with all device combinations.

vmoens commented on test/test_collector.py:
```diff
@@ -989,6 +989,83 @@ def test_collector_output_keys(collector_class, init_random_frames, explicit_spe
     del collector
 
 
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
```
You should probably use ˋget_available_devices` here as some tests run on machines that don't have a cuda device

vmoens commented on test/test_collector.py:
```diff
@@ -989,6 +989,83 @@ def test_collector_output_keys(collector_class, init_random_frames, explicit_spe
 
 
+@pytest.mark.parametrize("device", ["cuda", "cpu"])
+@pytest.mark.parametrize("passing_device", ["cuda", "cpu"])
```
Same as above

vmoens commented on test/test_collector.py:
```diff
@@ -989,6 +989,82 @@ def test_collector_output_keys(collector_class, init_random_frames, explicit_spe
     del collector
 
 
+@pytest.mark.parametrize(
```
Does it make sense to test on cpu only? Maybe we can skip it in that case

albertbou92 replied:
ok yes I can remove it in that case
